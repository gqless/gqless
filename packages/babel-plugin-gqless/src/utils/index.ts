export * from './importReferences'
export * from './findModule'
export * from './evaluate'
export * from './resolveRefInPattern'
export * from './methods'
