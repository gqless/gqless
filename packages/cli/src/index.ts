export * from './generate';
export * from './inspectWriteGenerate';
export * from './introspection';
export * from './writeGenerate';
export type { GQlessConfig as gqlessConfig, GQlessConfig } from './config';
export { defaultConfig, gqlessConfigPromise } from './config';
export { __innerState } from './innerState';
