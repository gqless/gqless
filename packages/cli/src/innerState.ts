export const __innerState = {
  isCLI: false,
};
