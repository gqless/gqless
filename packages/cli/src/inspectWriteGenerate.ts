import { existsSync } from 'fs';
import { promises } from 'fs';
import { buildSchema, GraphQLSchema, buildClientSchema } from 'graphql';
import { resolve } from 'path';

import { defaultConfig, DUMMY_ENDPOINT, gqlessConfigPromise } from './config';

import type { GenerateOptions, TransformSchemaOptions } from './generate';

export async function inspectWriteGenerate({
  endpoint,
  destination,
  generateOptions,
  cli,
  headers,
  transformSchemaOptions,
}: {
  /**
   * GraphQL API endpoint or GraphQL Schema file
   *
   * @example 'http://localhost:3000/graphql'
   * @example './schema.gql'
   */
  endpoint?: string;
  /**
   * File path destination
   * @example './src/gqless/index.ts'
   */
  destination?: string;
  /**
   * Specify generate options
   */
  generateOptions?: GenerateOptions;
  /**
   * Whether it's being called through the CLI
   */
  cli?: boolean;
  /**
   * Specify headers for the introspection HTTP request
   */
  headers?: Record<string, string>;
  /**
   * Specify schema transforms
   */
  transformSchemaOptions?: TransformSchemaOptions;
} = {}) {
  if (destination) {
    defaultConfig.destination = destination;
  }

  if (endpoint) {
    defaultConfig.introspection.endpoint = endpoint;
  } else if (existsSync(resolve('./schema.gql'))) {
    endpoint = './schema.gql';
    defaultConfig.introspection.endpoint = endpoint;
  } else {
    const { config, filepath } = await gqlessConfigPromise;

    const configIntrospectionEndpoint =
      config.introspection && config.introspection.endpoint;

    if (
      configIntrospectionEndpoint &&
      configIntrospectionEndpoint !== defaultConfig.introspection.endpoint
    ) {
      endpoint = configIntrospectionEndpoint;
    } else {
      console.error(
        `\nPlease modify "${
          filepath.endsWith('package.json') ? 'gqless' : 'config'
        }.introspection.endpoint" in: "${filepath}".`
      );
      throw Error(
        'ERROR: No introspection endpoint specified in configuration file.'
      );
    }
  }

  if (!destination) {
    const configDestination = (await gqlessConfigPromise).config.destination;

    destination = configDestination || defaultConfig.destination;
  }

  destination = resolve(destination);

  const genOptions = Object.assign({}, generateOptions);

  let schema: GraphQLSchema;

  defaultConfig.introspection.endpoint = endpoint;
  defaultConfig.introspection.headers = headers || {};

  if (endpoint.startsWith('http://') || endpoint.startsWith('https://')) {
    schema = await (await import('./introspection')).getRemoteSchema(endpoint, {
      headers,
    });
  } else {
    defaultConfig.introspection.endpoint = DUMMY_ENDPOINT;

    if (existsSync(endpoint)) {
      const file = await promises.readFile(endpoint, {
        encoding: 'utf-8',
      });

      if (endpoint.endsWith('.json')) {
        const parsedFile = JSON.parse(file);

        if (!parsedFile?.data?.__schema)
          throw Error(
            'Invalid JSON introspection result, expected "data.__schema" field.'
          );

        schema = buildClientSchema(parsedFile.data);
      } else {
        schema = buildSchema(file);
      }
    } else {
      throw Error(
        `File "${endpoint}" doesn't exists. If you meant to inspect a GraphQL API, make sure to put http:// or https:// in front of it.`
      );
    }
  }

  const generatedPath = await (await import('./writeGenerate')).writeGenerate(
    schema,
    destination,
    genOptions,
    async (existingFile) => {
      const subscriptions =
        genOptions.subscriptions ??
        (await gqlessConfigPromise).config.subscriptions;
      const react =
        genOptions.react ?? (await gqlessConfigPromise).config.react;

      const advice = `\nIf you meant to change this, please remove "${destination}" and re-run code generation.`;

      if (subscriptions) {
        if (!existingFile.includes('createSubscriptionsClient')) {
          console.warn(
            `[Warning] You've changed the option "subscriptions" to 'true', which is different from your existing "${destination}".` +
              advice
          );
        }
      }
      if (react) {
        if (!existingFile.includes('createReactClient')) {
          console.warn(
            `[Warning] You've changed the option "react" to 'true', which is different from your existing "${destination}".` +
              advice
          );
        }
      }
    },
    transformSchemaOptions
  );

  if (cli) {
    console.log('Code generated successfully at ' + generatedPath);
  }
}
