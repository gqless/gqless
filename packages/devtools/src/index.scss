@font-face {
  font-family: 'Operator mono';
  src: url(./fonts/OperatorMonoLig-Book.otf);
}

@font-face {
  font-family: 'Operator mono';
  src: url(./fonts/OperatorMonoLig-BookItalic.otf);
  font-style: italic;
}

@font-face {
  font-family: 'Operator mono';
  src: url(./fonts/OperatorMonoLig-Medium.otf);
  font-style: bold;
}

body {
  background: #ecf3ff;

  font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
    Ubuntu, Cantarell, Open Sans, Helvetica Neue, sans-serif;

  @media (prefers-color-scheme: dark) {
    background: #212c4f;
    color: #fff;
  }
}

*,
:after,
:before {
  box-sizing: inherit;
  font-family: inherit;
}
