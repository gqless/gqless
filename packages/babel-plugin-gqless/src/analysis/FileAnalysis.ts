import { dirname } from 'path'
import { types as t, NodePath, BabelFileResult, traverse } from '@babel/core'
import { Cache } from './Cache'
import { FunctionAnalysis } from './FunctionAnalysis'
import { Analysis } from './Analysis'
import {
  findModule,
  evalAsString,
  evalProperty,
  resolveRefInPattern,
} from '../utils'
import { invariant } from '@gqless/utils'

export type ImportSpecifier =
  | t.ImportSpecifier
  | t.ImportDefaultSpecifier
  | t.ImportNamespaceSpecifier

export class FileAnalysis extends Analysis {
  public exports = new Map<string, Analysis>()
  public functions = new Set<FunctionAnalysis>()

  private _ast?: t.File

  constructor(cache: Cache, public path: string) {
    super()
    this.file = this
    this.cache = cache
  }

  public getFunction(path: NodePath<t.Function>) {
    for (const funcAnalysis of this.functions) {
      if (funcAnalysis.path === path) return funcAnalysis
    }
    const funcAnalysis = new FunctionAnalysis(this, path)
    this.functions.add(funcAnalysis)
    return funcAnalysis
  }

  public get(
    path: NodePath<t.Node | null>,
    opts: { varName?: string; namespaceImport?: () => string } = {}
  ) {
    if (path.isFunction()) {
      return this.getFunction(path)
    }

    if (path.isIdentifier() || path.isMemberExpression()) {
      return this.getRef(path)
    }

    if (path.parentPath.isImportDeclaration()) {
      return this.getImport(
        path as NodePath<ImportSpecifier>,
        opts.namespaceImport
      )
    }

    if (path.isVariableDeclaration() || path.isVariableDeclarator()) {
      if (opts.varName === undefined) return
      return this.getVariable(path, opts.varName)
    }

    if (path.parentPath.isImportDeclaration()) {
      return this.getImport(path as NodePath<ImportSpecifier>)
    }

    return
  }

  public getImport(
    path: NodePath<ImportSpecifier>,
    namespaceImport?: () => string
  ) {
    const moduleAnalysis = this.getModule(
      (<t.ImportDeclaration>path.parent).source.value
    )

    if (path.isImportNamespaceSpecifier()) {
      const imp = namespaceImport?.()
      return imp ? moduleAnalysis.getExport(imp) : moduleAnalysis
    }

    const exportAnalysis = moduleAnalysis.getExport(
      (path.node as t.ImportSpecifier)?.imported?.name ?? 'default'
    )

    return exportAnalysis
  }

  public getVariable(
    path: NodePath<t.VariableDeclaration | t.VariableDeclarator>,
    name: string
  ): Analysis | void {
    if (path.isVariableDeclarator()) {
      const id = path.get('id')
      const init = path.get('init')
      if (init.node === null) return

      const value = resolveRefInPattern(id, init as NodePath, name)
      if (!value) return

      return this.get(value)
    }

    for (const declarator of (path as NodePath<t.VariableDeclaration>).get(
      'declarations'
    )) {
      const analysis = this.getVariable(declarator, name)
      if (analysis) return analysis
    }
  }

  public getRef(
    path: NodePath<t.Identifier | t.MemberExpression>,
    ...ctx: NodePath<t.MemberExpression>[]
  ): Analysis | void {
    if (path.isIdentifier()) {
      const binding = path.scope.getBinding(path.node.name)
      if (!binding) return

      return this.get(binding.path, {
        varName: path.node.name,
        namespaceImport() {
          return evalProperty(ctx[0])!
        },
      })
    }

    if (path.isMemberExpression()) {
      return this.getRef(path.get('object') as any, path, ...ctx)
    }
  }

  public getExport(name: string): Analysis | void {
    if (this.exports.has(name)) {
      return this.exports.get(name)
    }

    // Iterate over all top-level paths
    const findExport = (path: NodePath) => {
      if (path.isExportDefaultDeclaration()) {
        if (name !== 'default') return

        return this.get(path.get('declaration'))
      }

      // export var / { x } / * as x
      if (path.isExportNamedDeclaration()) {
        if (path.node.exportKind && path.node.exportKind !== 'value') return

        // export var x
        if (path.node.declaration) {
          const declaration = path.get('declaration')

          if (declaration.isFunctionDeclaration()) {
            if (declaration.node.id!.name !== name) return
            return this.getFunction(declaration)
          }

          if (declaration.isVariableDeclaration()) {
            return this.getVariable(declaration, name)
          }

          if (declaration.isClassDeclaration()) {
            if (declaration.node.id!.name !== name) return
            // TODO
          }
        }

        for (const specifier of path.get('specifiers')) {
          const moduleAnalysis = path.node.source
            ? this.file.getModule(evalAsString(path.get('source') as NodePath)!)
            : undefined

          // export { x }
          if (specifier.isExportSpecifier()) {
            if (specifier.node.exported.name !== name) continue
            const localName = specifier.node.local.name

            // export { x } from 'mod'
            if (moduleAnalysis) {
              return moduleAnalysis.getExport(localName)
            }

            const binding = specifier.scope.getBinding(localName)
            // exporting unknown variable
            if (!binding) return

            return this.get(binding.path, { varName: localName })
          }

          // export * as x from 'mod'
          else if (specifier.isExportNamespaceSpecifier()) {
            if (specifier.node.exported.name !== name) continue
            return moduleAnalysis!
          }

          // export default from 'mod'
          else if (specifier.isExportDefaultSpecifier()) {
            if (specifier.node.exported.name !== name) continue

            return moduleAnalysis!.getExport('default')
          }
        }
      }

      // export * from
      if (path.isExportAllDeclaration()) {
        const moduleAnalysis = this.file.getModule(
          evalAsString(path.get('source') as NodePath)!
        )

        return moduleAnalysis.getExport(name)
      }
    }

    for (const path of this.programPath.get('body')) {
      const exportAnalysis = findExport(path)
      if (!exportAnalysis) continue

      this.exports.set(name, exportAnalysis)
      return exportAnalysis
    }
  }

  public getModule(id: string) {
    try {
      const modulePath = findModule(dirname(this.path), id)
      return this.cache.getPath(modulePath)
    } catch (e) {
      throw new Error(
        `Failed resolving module relative to ${this.file.path}. \n${e.message}`
      )
    }
  }

  private _programPath: NodePath<t.Program> | undefined
  public get programPath() {
    if (!this._programPath) {
      traverse(this.ast, {
        Program: path => {
          this._programPath = path
        },
      })
    }

    return this._programPath!
  }

  public set ast(ast: t.File) {
    this._ast = ast
  }

  public get ast() {
    if (!this._ast) {
      this.ast = this.cache.transform(this.path)
    }

    return this._ast!
  }
}
