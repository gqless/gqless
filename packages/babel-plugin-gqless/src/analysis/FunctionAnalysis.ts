import { types as t, NodePath } from '@babel/core'
import { Analysis } from './Analysis'
import { ParamAnalysis } from './ParamAnalysis'
import { Referable } from './mixins'
import { Mix, Generic as g } from 'mix-classes'

export interface FunctionAnalysis extends Referable<t.Function> {}

export class FunctionAnalysis extends Mix(Analysis, g(Referable)) {
  public params = new Map<number | null, ParamAnalysis>()

  constructor(parent: Analysis, path: NodePath<t.Function>) {
    super([parent], [path])
  }

  public getParam(index: number) {
    const params = this.path.get('params')

    if (index >= params.length) {
      const lastParam = params[params.length - 1]
      if (!lastParam?.isSpreadElement()) return

      if (this.params.has(null)) {
        return this.params.get(null)!
      }

      const paramAnalysis = new ParamAnalysis(this, lastParam)
      this.params.set(null, paramAnalysis)
      return paramAnalysis
    }

    const param = params[index]

    for (const [i, paramAnalysis] of this.params) {
      if (i === index) return paramAnalysis
    }

    const paramAnalysis = new ParamAnalysis(this, param)
    this.params.set(index, paramAnalysis)
    return paramAnalysis
  }

  public scan(...args: NodePath<t.Expression | t.SpreadElement>[]) {
    const params = this.path.get('params')

    // Iterate over all params
    for (let i = 0; i < params.length; i++) {
      const param = params[i]

      if (param.isSpreadElement()) {
        const spreadArgs = args.slice(-i)
        if (!spreadArgs.length) break
      } else {
        const arg = args[i]
        if (!arg) break
      }

      this.getParam(i)?.scan()
    }

    if (this.parent !== this.parent!.file) {
      console.log('add analysis to prev analysis')
    }
  }
}
