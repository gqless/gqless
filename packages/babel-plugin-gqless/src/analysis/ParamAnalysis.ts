import { types as t, NodePath } from '@babel/core'
import { FunctionAnalysis } from './FunctionAnalysis'
import { Fields, Referable } from './mixins'
import { Analysis } from './Analysis'
import { Mix, Generic as g } from 'mix-classes'
import { objectPropValue, evalProperty } from '../utils'

type Param = t.Identifier | t.Pattern | t.RestElement | t.TSParameterProperty

export interface ParamAnalysis extends Referable<Param> {}

export class ParamAnalysis extends Mix(Analysis, g(Referable), Fields) {
  constructor(parent: FunctionAnalysis, path: NodePath<Param>) {
    super([parent], [path])
  }

  public scan() {
    this.scanPath(this.path)
  }
}
