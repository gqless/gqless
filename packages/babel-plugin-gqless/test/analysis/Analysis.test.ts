import { fileAnalysis, scan } from '../setup'
import { FunctionAnalysis, FileAnalysis } from '../../src/analysis'

describe('resolves', () => {
  describe('exports >', () => {
    test('export default', () => {
      const analysis = fileAnalysis({
        a: `
          export default () => {}
        `,
      })

      expect(analysis.getExport('default')).toBeInstanceOf(FunctionAnalysis)
    })

    test('export {x}', () => {
      const analysis = fileAnalysis({
        a: `
          const x = () => {}
          export { x }
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FunctionAnalysis)
    })

    test('export const x', () => {
      const analysis = fileAnalysis({
        a: `
          export const a = () => {}
        `,
      })

      expect(analysis.getExport('a')).toBeInstanceOf(FunctionAnalysis)
    })
  })

  describe('imports >', () => {
    test('import {x}', () => {
      const analysis = fileAnalysis({
        a: `
          import { f as b } from 'b'
          export const f = b
        `,
        b: `
          export const f = () => {}
        `,
      })

      expect(analysis.getExport('f')).toBeInstanceOf(FunctionAnalysis)
    })

    test('import x', () => {
      const analysis = fileAnalysis({
        a: `
          import f from 'b'
          export const b = f
        `,
        b: `
          export default () => {}
        `,
      })

      expect(analysis.getExport('b')).toBeInstanceOf(FunctionAnalysis)
    })

    test('import * as x', () => {
      const analysis = fileAnalysis({
        a: `
          import * as b from 'b'
          export const f = b.f
        `,
        b: `
          export const f = () => {}
        `,
      })

      expect(analysis.getExport('f')).toBeInstanceOf(FunctionAnalysis)
    })
  })

  describe('re-exports >', () => {
    test('export {x} from', () => {
      const analysis = fileAnalysis({
        a: `
          export { f as x } from 'b'
        `,
        b: `
          export const f = () => {}
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FunctionAnalysis)
    })

    test('export * from', () => {
      const analysis = fileAnalysis({
        a: `
          export * from 'b'
        `,
        b: `
          export const x = () => {}
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FunctionAnalysis)
    })

    test('export x from', () => {
      const analysis = fileAnalysis({
        a: `
          export x from 'b'
        `,
        b: `
          export default () => {}
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FunctionAnalysis)
    })

    test('export * as x from', () => {
      const analysis = fileAnalysis({
        a: `
          export * as x from 'b'
        `,
        b: `
          export const f = () => {}
        `,
      })

      expect(analysis.getExport('x')).toBeInstanceOf(FileAnalysis)
    })
  })
})

describe('evaluates values', () => {
  test('with recursive de-structuring', () => {
    const analysis = fileAnalysis({
      a: `
        const { b } = { b: { a: () => {} } }

        export const { a } = b
      `,
    })

    expect(analysis.getExport('a')).toBeInstanceOf(FunctionAnalysis)
  })

  test('with object de-structuring', () => {
    const analysis = fileAnalysis({
      a: `
        const n = 'n'
        const fn = () => {}

        export const { ['f'+n]: a } = { fn }
      `,
    })

    expect(analysis.getExport('a')).toBeInstanceOf(FunctionAnalysis)
  })

  test('with array de-structuring', () => {
    const analysis = fileAnalysis({
      a: `
        export const [, a] = [, () => {}]
      `,
    })

    expect(analysis.getExport('a')).toBeInstanceOf(FunctionAnalysis)
  })
})

describe('scans', () => {
  test('arg destructuring', () => {
    const files = {
      a: `
        export default ({ user }) => {
          user.name
        }
      `,
    }

    expect(scan(files, 'props')).toMatchSnapshot()
  })

  test('field variables', () => {
    const files = {
      a: `
        export default u => {
          u.avatarUrl({ size: 100 })
        }
      `,
    }

    expect(scan(files, 'props')).toMatchInlineSnapshot(`
      FunctionAnalysis (
        0 -> {
          avatarUrl({
            "size": 100
          }) {}
        }
      )
    `)
  })

  describe('imports >', () => {
    const scanImp = (source: string) =>
      scan(
        {
          a: source,
          b: `
              export const b = u => u.n
              export default b
            `,
        },
        'props'
      )

    test('import x', () => {
      expect(
        scanImp(`
          import x from 'b'
          export default x
        `)
      ).toMatchSnapshot()
    })

    test('import {x}', () => {
      expect(
        scanImp(`
          import { b as _ } from 'b'
          export default _
        `)
      ).toMatchSnapshot()
    })

    test('import * as x', () => {
      expect(
        scanImp(`
          import * as b from 'b'
          export default b.default
        `)
      ).toMatchSnapshot()
    })
  })

  describe('variables', () => {
    test('with destructuring', () => {
      const files = {
        a: `
          export default u => {
            const { a: { b, ...a }, ...o } = { a: u.a }

            o.a.b3
            a.b2
            b.c
          }
        `,
      }

      expect(scan(files, 'props')).toMatchSnapshot()
    })

    test('with spread', () => {
      const files = {
        a: `
            export default u => {
              const _u = { ...u }

              _u.a
            }
          `,
      }

      // expect(scan(files, 'props')).toMatchInlineSnapshot(`
      //   FunctionAnalysis (
      //     0 -> {
      //       a {}
      //     }
      //   )
      // `)
    })
  })

  describe('function calls', () => {
    test('with object arguments', () => {
      const files = {
        a: `
          const a = ({ u }) => u.age
          export default u => {
            a({ u })
            u.name
          }
        `,
      }

      expect(scan(files, 'props')).toMatchSnapshot()
    })

    test('with multiple arguments', () => {
      const files = {
        a: `
          const a = (a, b, c) => {
            a.a
            b.b
            c.c
          }

          export default u => {
            a(u, u, {})
            u.name
          }
        `,
      }

      expect(scan(files, 'props')).toMatchSnapshot()
    })
  })

  describe('array', () => {
    test('methods', () => {
      const files = {
        a: `
          export default u => {
            u.filter(u => u.a).map(u => u.b)
            u.forEach((u, i, _u) => {
              _u.x
              u.c
            })
            u.find(u => u.d).e
          }
        `,
      }

      expect(scan(files, 'props')).toMatchSnapshot()
    })

    test('for of', () => {
      const files = {
        a: `
          export default u => {
            for (const x of u) {
              x.a
            }

            let y
            for (y of u) {
              y.b
            }
          }
        `,
      }

      expect(scan(files, 'props')).toMatchSnapshot()
    })

    test('for loop', () => {
      const files = {
        a: `
          export default u => {
            for (let i = 0; i < u.length; i++) {
              u[i].a
            }
          }
        `,
      }

      expect(scan(files, 'props')).toMatchSnapshot()
    })
  })
})
