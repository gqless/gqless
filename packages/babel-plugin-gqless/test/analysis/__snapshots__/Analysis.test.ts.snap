// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scans arg destructuring 1`] = `
FunctionAnalysis (
  0 -> {
    user {
      name {}
    }
  }
)
`;

exports[`scans array for loop 1`] = `
FunctionAnalysis (
  0 -> {
    length {}
    INDEX {
      a {}
    }
  }
)
`;

exports[`scans array for of 1`] = `
FunctionAnalysis (
  0 -> {
    INDEX {
      a {}
      b {}
    }
  }
)
`;

exports[`scans array methods 1`] = `
FunctionAnalysis (
  0 -> {
    INDEX {
      a {}
      b {}
      c {}
      d {}
      e {}
    }
    x {}
  }
)
`;

exports[`scans function calls with multiple arguments 1`] = `
FunctionAnalysis (
  0 -> {
    a {}
    b {}
    name {}
  }
)
`;

exports[`scans function calls with object arguments 1`] = `
FunctionAnalysis (
  0 -> {
    age {}
    name {}
  }
)
`;

exports[`scans imports > import * as x 1`] = `
FunctionAnalysis (
  0 -> {
    n {}
  }
)
`;

exports[`scans imports > import {x} 1`] = `
FunctionAnalysis (
  0 -> {
    n {}
  }
)
`;

exports[`scans imports > import x 1`] = `
FunctionAnalysis (
  0 -> {
    n {}
  }
)
`;

exports[`scans variables with destructuring 1`] = `
FunctionAnalysis (
  0 -> {
    a {
      b {
        c {}
      }
      b2 {}
      b3 {}
    }
  }
)
`;
